packages/engine/lib/vehicle.ex

```diff
@@ -1,113 +1,104 @@
+    |> Booking.assign()
+  end
 
-  defp do_get_indices(pickup_index, _, list_length) when pickup_index >= list_length,
-    do: []
+  def handle_driver_response({:ok, false}, _, _), do: IO.puts("Driver didnt want the booking :(")
 
-  defp do_get_indices(pickup_index, delivery_index, list_length)
-       when delivery_index == list_length,
-       do: do_get_indices(pickup_index + 1, pickup_index + 2, list_length)
+  def make(external_id, position, busy \\ false) do
+    id = external_id
 
-  defp do_get_indices(pickup_index, delivery_index, list_length),
-    do:
-      do_get_indices(pickup_index, delivery_index + 1, list_length)
-      |> Enum.concat([{pickup_index, delivery_index}])
+    GenServer.start_link(
+      __MODULE__,
+      %Vehicle{
+        id: id,
+        external_id: external_id,
+        position: position,
+        busy: busy
+      },
+      name: via_tuple(id)
+    )
 
-  def get_osrm_distance(instructions) do
-    Osrm.route(instructions)
-    |> Map.get(:distance)
+    id
   end
 
-  def offer(%Vehicle{id: id, instructions: instructions, booking_ids: booking_ids} = vehicle) do
-    IO.inspect(vehicle, label: "offer to vehicle")
-    instructions_without_start = Enum.filter(instructions, &Map.has_key?(&1, :id))
+  defp via_tuple(id) when is_binary(id), do: via_tuple(String.to_integer(id))
 
-    accepted_bookings =
-      booking_ids
-      |> Enum.map(fn id ->
-        Enum.filter(instructions_without_start, fn instruction -> instruction.id == id end)
-      end)
-      |> IO.inspect(label: "instructions for booking")
-      |> Enum.map(fn instructions_for_booking ->
-        instruction =
-          instructions_for_booking
-          |> Enum.map(fn %{address: address, type: type} -> {address, type} end)
-
-        booking =
-          Map.new()
-          |> Map.put(
-            :pickup,
-            Enum.find(instruction, fn {_, type} -> type == "pickupShipment" end)
-            |> elem(0)
-          )
-          |> Map.put(
-            :delivery,
-            Enum.find(instruction, fn {_, type} -> type == "deliverShipment" end)
-            |> elem(0)
-          )
-          |> IO.inspect(label: "booking")
-
-        {:ok, accepted} =
-          MQ.call(%{vehicle: %{id: id}, booking: booking}, "pickup_offers", "p_response")
-          |> Poison.decode()
-          |> IO.inspect(label: "the driver answered")
-
-        %{vehicle: vehicle, booking: booking, accepted: accepted}
-      end)
-      |> Enum.filter(fn %{accepted: accepted} -> accepted end)
-
-    %{vehicle: vehicle, booking_ids: booking_ids, accepted_bookings: accepted_bookings}
+  defp via_tuple(id) when is_integer(id) do
+    IO.inspect(id, label: "via tuple")
+    {:via, :gproc, {:n, :l, {:vehicle_id, id}}}
+  end
+
+  def get(id) do
+    GenServer.call(via_tuple(id), :get)
+  end
+
+  def offer(%Vehicle{id: id, instructions: instructions, booking_ids: booking_ids} = vehicle) do
+    GenServer.cast(via_tuple(id), {:offer, vehicle})
```

Comment:
Looks good! But in the future we should probably derive as much as we can from state rather than sending in a whole vehicle to the "offer"-call. The only thing we need is the vehicle Id to find the appropriate vehicle process, then we can get everything else from that process' state. This we will cause less weirdness in the future :)